fix: Improved error handling

tombh opened the pull request:
This includes a notable, perhaps even controversial, behaviour change:
All errors other than LSP requests are now sent to the client as `showMessage.type = MessageType.Error` messages. This may result in existing custom LSP servers showing errors that were not seen before, these aren't new errors, just previously undisplayed errors.

Another advantage to this is that it is now easier to test error handling. When running end to end tests, all errors are now accessible in the client.

Also, `JsonRPCProtocol.data_receieved()` now catches unhandled errors.


Fixes #277.

- [x] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR)
- [x] Title summarizes what is changing
- [x] Commit messages are meaningful (see [this][commit messages] for details)
- [x] Tests have been included and/or updated, as appropriate
- [x] Docstrings have been included and/or updated, as appropriate
- [x] Standalone docs have been updated accordingly
- [x] [CONTRIBUTORS.md][contributors] was updated, as appropriate
- [x] Changelog has been updated, as needed (see [CHANGELOG.md][changelog])

[changelog]: https://github.com/openlawlibrary/pygls/blob/master/CHANGELOG.md
[commit messages]: https://chris.beams.io/posts/git-commit/
[contributors]: https://github.com/openlawlibrary/pygls/blob/master/CONTRIBUTORS.md

perrinjerome commented on pygls/protocol.py:
```diff
@@ -392,8 +396,14 @@ def _send_data(self, data):
                 self.transport.write(header + body)
             else:
                 self.transport.write(body.decode('utf-8'))
+            print(data)
```
```suggestion
```

tombh replied:
Deleted, thanks

perrinjerome commented on docs/source/pages/advanced_usage.rst:
```diff
@@ -447,6 +447,19 @@ And method invocation example:
+.. code:: python
+
+    Class CustomLanguageServer(LanguageServer):
+        def report_server_error(self, message: str):
```
how about a signature like this ?
 
```py
        def report_server_error(self, error: Exception):
```

then custom server implementations would have more context.

alcarney replied:
Seconded :)

tombh replied:
Thirded!

perrinjerome commented on pygls/protocol.py:
```diff
@@ -427,6 +437,17 @@ def connection_made(self, transport: asyncio.BaseTransport):
+        try:
+            self._data_received(data)
+        except Exception:
+            logger.error(traceback.format_exc())
```
I know that this pattern is present in other places, but instead of:
```py
            logger.error(traceback.format_exc())
```
we can use logging ability to log the currently handled exception with something like this:

```py
            logger.error("Error receiving data", exc_info=True)
```

tombh replied:
Updated in latest force push

alcarney commented:
Even though there is a mechanism for sending an error response for JSON-RPC requests, do you think it's worth calling `report_server_error` anyway? It feels odd to have a function that's called for any error.... except this one scenario.

It might also be worth adding a second argument to `report_server_error` that tells it which situation it's being called in.
I imagine the default implementation would want to handle all cases apart from requests.

tombh commented:
I've updated the commit:
* Replaced `message: str` arg with `error: Exception` arg in `report_server_error()`. But still not sending that error to the editor by default (a plain human-readable message is sent instead). So its use case is more for authors that want to override the default `report_server_error()`.
* `report_server_error()` now takes a `source` arg so that request errors can also be potentially dual-reported, though their sending as `showMessage` errors is suppressed by default (current behaviour of including any errors in the LSP response object remains unchanged). At first I thought an enum would be good. Then I thought custom errors, which is a bit weird because you'd think that you could get that custom error from the `error` arg, but of course these are unexpected errors so the underlying exception type could be anything. So I've created these 2 new error types anyway: `FeatureRequestError(PyglsError)` and `FeatureNotificationError(PyglsError)`. Which is indeed a little curious because they're never raised, they're only used for matching the error source. But it feels better to create them rather than enums, because at least these can have other use cases the code base.
* Also some other refactoring. And a test to check that custom `report_server_error()` methods can inadvertently create recursive error reports.

alcarney commented on pygls/server.py:
```diff
@@ -426,6 +433,37 @@ def show_message_log(self, message, msg_type=MessageType.Log) -> None:
+        default.
+        """
+
+        if isinstance(source, FeatureRequestError):
```
Since the error classes rather than instances are being passed to `source`, I believe this check should be `issubclass` instead

alcarney commented:
Made a suggestion on the if check in the default `report_server_error` implementation, but otherwise looking good!

tombh commented:
Oh! Yes, I should have known that 🤦 And in fact, the test that was testing for that was a false positive, so I wrote an extra test that tests that test hahah

tombh commented:
Oh, actually, saying that, Mypy wasn't happy and actually `if source == FeatureRequestError:` seems to work. Make sense?

tombh commented:
Awesome, thanks for the feedback everybody. LGTM 🚀. Just added the CHANGELOG deets, can I get another review acceptance please 🙇
